### ISSUE TITLE
"novalidate" gets ignored if Formula is used

### ISSUE THREAD

**Issue by sserdyuk** [tanepiper/svelte-formula#7]
Hi,

First of all, thank you for starting this package. I've evaluated a number of options, and liked yours better. I am putting together a small project to see if Svelte can be "enterprise grade" solution, and reactive forms are a big part of it.

I ran across a weird issue that I have not figured out a fix for, but I wanted to share in case it immediately makes sense to you.
Since validation is handled by Formula, I want to avoid browser's built in validation handling, and so I put `novalidate` attribute on the form. As soon as I decorated it with Formula using `use:form`, the built in validation start appearing. I am including a screenshot of the undesired validation message and the repl it can be observed on.

https://svelte.dev/repl/ea846ec78fd34ab9bde899007d546005?version=3.35.0

![image](https://user-images.githubusercontent.com/395998/110728938-dbcee880-81eb-11eb-9e2a-68a28fbd9639.png)

**Comment by tanepiper** [comment]
Hi @sserdyuk thanks for reporting - good call, while I was aware of it I hadn't tested the `novalidate` situation (I personally haven't needed to use it) but it'll be a good test case later! Let me have a look into it and get back to you what the outcome is.

**Comment by tanepiper** [comment]
@sserdyuk OK quick look I think it was fixable - in this case I check for the attribute and pass it down into the instance where if it's true then all validations will return true (so we disable both HTML and additional validations passed to the constructor)

**Comment by sserdyuk** [comment]
Just to be clear, I don't want to disable Formula validations. I just want
to get rid of browser's default validations.

--Regards, Sergei. / sent from my phone, please excuse typos /

**Comment by tanepiper** [comment]
@sserdyuk Ahh ok and yes it makes sense I suppose - let me have a think about it because I suspect it's where I'm calling `el.checkValidity()` - I wonder if doing this even if `novalidate` is set it causing it because otherwise, I'm only reading the validation state from an HTML5 set one (then custom validations are applied)

Edit: Yep - quick check here - https://svelte.dev/repl/2fcaae631d664dbc8d61744b4679b222?version=3.35.0

Even if `novalidate` is set, `el.checkValidity()` *still* returns true or false - so here it looks like we just need to ignore this *if* novalidate it set.  I'll try have something sensible out later today

**Comment by tanepiper** [comment]
@sserdyuk OK I want to clarify your use case here by explaining the code here:

If there are no custom options it immediately returns the browser validity - so here if there is `novalidate` it's easy to skip.

Then custom validations are run - here if there are any they are set on the element itself using `el.setCustomValidity` then validation is called again. 

```ts
// Check for any custom messages in the options or dataset
const customMessages = { ...options?.messages?.[inputGroup], ...el.dataset };
// Check for any custom validations
const [messages, customErrors] = getCustomValidations(elValue, options?.validators?.[inputGroup]);

const errors = extractErrors(el, customErrors);
const errorKeys = Object.keys(errors);
// If there is no field validity issues, set custom ones
if (el.checkValidity()) {
  if (errorKeys.length > 0) {
    el.setCustomValidity(messages[errorKeys[0]]);
  }
  // Check for custom messages
} else {
  if (customMessages[errorKeys[0]]) {
    el.setCustomValidity(customMessages[errorKeys[0]]);
  }
}
// Recheck validity and show any messages
const valid = el.checkValidity();
if (!valid) {
  el.setAttribute('data-formula-invalid', 'true');
}
```

If I get what your suggesting it's that the entire block after `getCustomValidations` is not run and only return errors from that validator?

**Comment by sserdyuk** [comment]
@tanepiper  I fill like my intent didn't come across. Let me give you a bit of context. 

I like how Formula uses standard HTML5  validation attributes on form elements to express the requirements. I do prefer the brevity of this approach to coding validations at the `formula()` call. This works beautifully for simple forms where validations are not conditional.  

At the same time, I do not like browser's default validation UI. It is inconsistent from one browser to another and also (arguably) ugly. I want to be able to use the `validity` store to trigger UI into a "validation failed" state and display the messages. This way it can be precisely controlled. 

So, basically, I want the validations to run regardless `novalidate` attribute's presence or value. I do want to be able to trigger the form into the "invalid" state. All I want to get rid of is that ugly popup that browsers display if I use HTML5 validation attributes on a form. The standard way to prevent browser's default UI to appear is to put the `novalidate` on the form. That's the reason I put it on the form.

Now, let me restate the problem. As soon as I hook up the form using `use:form` the `novalidate` attribute no longer prevents browser's default UI from appearing.

**Comment by tanepiper** [comment]
@sserdyuk OK, so I found the issue - I had recently added a call to the form validity method if it's a form element, and this was triggering it.  It now checks for novalidate first - just published 0.9.3 if you can check it

**Comment by sserdyuk** [comment]
@tanepiper Thank you. I can confirm that the issue was resolved.
